aspnet/fundamentals/app-state/sample/src/AppState/Startup.cs

```diff
@@ -0,0 +1,128 @@
+            if (env.IsDevelopment())
+            {
+                app.UseDeveloperExceptionPage();
+            }
+
+            // don't count favicon requests
+            app.Map("/favicon.ico", ignore => { });
+
+            // example middleware that does not reference session at all and is configured before app.UseSession()
+            app.Map("/untracked", subApp =>
+            {
+                subApp.Run(async context =>
+                {
+                    await context.Response.WriteAsync("<html><body>");
+                    await context.Response.WriteAsync("Requested at: " + DateTime.Now.ToLongTimeString() + "<br>");
+                    await context.Response.WriteAsync("This part of the application isn't referencing Session...<br><a href=\"/\">Return</a>");
+                    await context.Response.WriteAsync("</body></html>");
+                });
+            });
+
+            app.UseSession();
+
+            // establish session
+            app.Map("/session", subApp =>
+            {
+                subApp.Run(async context =>
+                {
+                    RequestEntryCollection collection = GetOrCreateEntries(context);
+                    collection.Add(context.Request.PathBase + context.Request.Path);
+                    SaveEntries(context, collection);
+                    context.Session.SetString("StartTime", DateTime.Now.ToLongTimeString());
+
+                    await context.Response.WriteAsync("<html><body>");
+                    await context.Response.WriteAsync($"Counting: You have made {collection.TotalCount()} requests to this application.<br><a href=\"/\">Return</a>");
+                    await context.Response.WriteAsync("</body></html>");
+
+                });
+            });
+
+            // main catchall middleware
+            app.Run(async context =>
+            {
+                RequestEntryCollection collection = GetOrCreateEntries(context);
+
+                await context.Response.WriteAsync("<html><body>");
+                if (collection.TotalCount() == 0)
+                {
+                    await context.Response.WriteAsync("Your session has not been established.<br>");
+                    await context.Response.WriteAsync(DateTime.Now.ToLongTimeString() + "<br>");
+                    await context.Response.WriteAsync("<a href=\"/session\">Establish session</a>.<br>");
+                }
+                else
+                {
+                    collection.Add(context.Request.PathBase + context.Request.Path);
+                    await context.Response.WriteAsync("Session Established At: " + context.Session.GetString("StartTime") + "<br>");
+                    foreach (var entry in collection.Entries)
+                    {
+                        await context.Response.WriteAsync("Request: " + entry.Path + " was requested " + entry.Count + " times.<br />");
+                    }
+                    SaveEntries(context, collection);
```

Comment:
It's best to consistently do your session operations  before writing to the response. In this case it won't matter because you've already verified the session has been established, but if this was re-used elsewhere it could fail.